File: kmp/shared-core/spm/Sources/SharedCoreKit/ReportReason.swift
import Foundation
import SharedCore

/// The Kotlin vocabulary, under a local name. Deliberately a typealias and not a Swift enum:
/// the tokens are the wire format, and there must be exactly one place they are written down.
public typealias ReportReason = SharedCore.ReportReason

/// Assembles the `description` field of a `ReportRequest`, in the shared Kotlin.
public enum ReportDescription {

    /// The cap the composer enforces as you type.
    public static let maxDetailsLength = Int(SharedCore.ReportDescription.shared.MAX_DETAILS_LENGTH)

    /// The `description` field of a `ReportRequest`: reason token, then the person's words.
    public static func build(reason: ReportReason, details: String?) -> String {
        SharedCore.ReportDescription.shared.build(reason: reason, details: details)
    }
}

public extension ReportReason {
    /// The order the reasons are offered in. This *is* app UI, which is why it lives on the
    /// Swift side -- but it must stay exhaustive, and `ReportReasonTests` checks that it is.
    ///
    /// The case names are the ones Kotlin/Native generates, which lowercases the whole word
    /// rather than preserving the Kotlin hump: `scamorfraud`, not `scamOrFraud`.
    static var displayOrder: [ReportReason] {
        [.spam, .scamorfraud, .harassment, .sexualcontent, .violence, .other]
    }
}

File: libs/reporting/src/commonMain/kotlin/com/flipcash/reporting/ReportReason.kt
package com.flipcash.reporting

/**
 * The reason a person gave for a report, and the token that carries it to the server.
 *
 * The contract (`flipcash2-client-protocol` 0.10.0) has no category field -- `ReportRequest`
 * carries only a free-form `description`. These tokens are how a category survives that,
 * so they are part of the wire format: renaming one changes what moderation receives, and
 * `ReportDescriptionTest` exists to make that change fail loudly rather than silently.
 *
 * Labels, ordering and icons are deliberately not here. Those are per-app UI.
 */
enum class ReportReason(val token: String) {
    Spam("spam"),
    ScamOrFraud("scam_or_fraud"),
    Harassment("harassment"),
    SexualContent("sexual_content"),
    Violence("violence"),
    Other("other"),
}
